envoy/network/transport_socket.h

```diff
@@ -230,6 +230,22 @@ class TransportSocketOptions {
    */
   virtual absl::optional<Network::ProxyProtocolData> proxyProtocolOptions() const PURE;
 
+  // Information for use by the http_11_proxy transport socket.
+  struct ProxyInfo {
+    ProxyInfo(std::string hostname, Network::Address::InstanceConstSharedPtr address)
+        : hostname(hostname), proxy_address(address) {}
+    // The hostname of the original request, to be used in CONNECT request if
+    // the underlying transport is TLS.
+    std::string hostname;
+    // The address of the proxy, where connections should be routed to.
+    Network::Address::InstanceConstSharedPtr proxy_address;
+  };
+
+  /**
+   * @return any proxy information if sending to an intermediate proxy over HTTP/1.1.
+   */
+  virtual const std::unique_ptr<const ProxyInfo>& proxyInfo() const PURE;
```

Comment:
Why not keep this private to the extension? E.g. per https://github.com/envoyproxy/envoy/pull/21700, would be another private hashable object.

Reply:
These objects are also hashed in transport options and passed along during connection creation to the upstream connection stream info. Maybe I'm missing something?

Reply:
I see how I could handle hashing but I don't see how I could handle the createConnection in  
source/common/upstream/upstream_impl.cc properly without explicit visibility of the addresses?

Reply:
I see. I suppose we could split the two since `address()` override seems more generic than CONNECT (you could do passthrough and all sorts of things with it), but it's up to you.